### ISSUE TITLE
bootstrap linux mixes crt1 from bootstrap libc and libc.so.6 from new libc, sometimes fails linking

### ISSUE THREAD

**Issue by trofi** [NixOS/nixpkgs#158042]
The best way to show the problem is to try to build current nixpkgs against glibc.2.35, patch:
- ~~[0001-glibc-2.33-2.35.patch.txt](https://github.com/NixOS/nixpkgs/files/7998198/0001-glibc-2.33-2.35.patch.txt)~~ 
- ~~[v2-0001-glibc-2.33-2.35.patch.txt](https://github.com/NixOS/nixpkgs/files/8052200/v2-0001-glibc-2.33-2.35.patch.txt)~~
- [v3-0001-glibc-2.33-2.35.patch.txt](https://github.com/NixOS/nixpkgs/files/8052388/v3-0001-glibc-2.33-2.35.patch.txt)

It fails to build simplest program as:

```
$ nix build -f. lv -L --max-jobs 1
expand-response-params> unpacking sources
expand-response-params> patching sources
expand-response-params> configuring
expand-response-params> no configure script, doing nothing
expand-response-params> building
expand-response-params> /nix/store/w2x0imn84ckw6y83p72rigj3h5ymmwll-binutils-2.35.2/bin/ld: /nix/store/p4s4jf7aq6v6z9iazll1aiqwb34aqxq9-bootstrap-tools/lib/crt1.o: in function `_start':
expand-response-params> /build/glibc-2.27/csu/../sysdeps/x86_64/start.S:101: undefined reference to `__libc_csu_fini'
expand-response-params> /nix/store/w2x0imn84ckw6y83p72rigj3h5ymmwll-binutils-2.35.2/bin/ld: /build/glibc-2.27/csu/../sysdeps/x86_64/start.S:102: undefined reference to `__libc_csu_init'
expand-response-params> collect2: error: ld returned 1 exit status
```

If I add a bit of debugging it becomes more clear where mismatch comes from:

```diff
--- a/pkgs/build-support/expand-response-params/default.nix
+++ b/pkgs/build-support/expand-response-params/default.nix
@@ -10,7 +10,8 @@ stdenv.mkDerivation {
     src=$PWD
   '';
   buildPhase = ''
-    NIX_CC_USE_RESPONSE_FILE=0 "$CC" -std=c99 -O3 -o "expand-response-params" expand-response-params.c
+    export NIX_DEBUG=1
+    NIX_CC_USE_RESPONSE_FILE=0 "$CC" -std=c99 -O3 -o "expand-response-params" expand-response-params.c -Wl,--verbose
   '';
   installPhase = ''
     mkdir -p $prefix/bin
```

```
$ nb --max-jobs 1 -L lv |& grep -P '(crt1|libc[.]so).*succeed'
expand-response-params> attempt to open /nix/store/i3ibpx67yncp4w4mpkf5pwvjjsd0aqln-bootstrap-tools/lib/crt1.o succeeded
expand-response-params> attempt to open /nix/store/bf7l4af8rrfy1s17af4j2az0khlp5rw0-glibc-2.35/lib/libc.so succeeded
expand-response-params> attempt to open /nix/store/bf7l4af8rrfy1s17af4j2az0khlp5rw0-glibc-2.35/lib/libc.so.6 succeeded
```

My theory is that it was not visible before because until `glibc-2.35` `crt1.o` did not change much. But upstream https://sourceware.org/git/?p=glibc.git;a=commitdiff;h=035c012e32c11e84d64905efaf55e74f704d3668 change made it visible.

**Comment by trofi** [comment]
I think it happens mechanically because `-B` and `-L` options go in wrong order:

- `-B` goes "old libc", then "new new libc"
- `-L` goes "new libc", then "old libc"

```
expand-response-params> extra flags before to /nix/store/i3ibpx67yncp4w4mpkf5pwvjjsd0aqln-bootstrap-tools/bin/gcc:
expand-response-params>   -O2
expand-response-params>   -D_FORTIFY_SOURCE=2
expand-response-params>   -fstack-protector-strong
expand-response-params>   --param
expand-response-params>   ssp-buffer-size=4
expand-response-params>   -fno-strict-overflow
expand-response-params>   -Wformat
expand-response-params>   -Wformat-security
expand-response-params>   -Werror=format-security
expand-response-params>   -fPIC
expand-response-params>   -Wl\,-dynamic-linker=/nix/store/bf7l4af8rrfy1s17af4j2az0khlp5rw0-glibc-2.35/lib/ld-linux-x86-64.so.2
expand-response-params> original flags to /nix/store/i3ibpx67yncp4w4mpkf5pwvjjsd0aqln-bootstrap-tools/bin/gcc:
expand-response-params>   -std=c99
expand-response-params>   -O3
expand-response-params>   -o
expand-response-params>   expand-response-params
expand-response-params>   expand-response-params.c
expand-response-params>   -Wl\,--verbose
expand-response-params> extra flags after to /nix/store/i3ibpx67yncp4w4mpkf5pwvjjsd0aqln-bootstrap-tools/bin/gcc:
expand-response-params>   -B/nix/store/i3ibpx67yncp4w4mpkf5pwvjjsd0aqln-bootstrap-tools/lib
expand-response-params>   -B/nix/store/bf7l4af8rrfy1s17af4j2az0khlp5rw0-glibc-2.35/lib/
expand-response-params>   -idirafter
expand-response-params>   /nix/store/3cmdhd36l2cqm484l381hz8z0nmq35fy-glibc-2.35-dev/include
expand-response-params>   -idirafter
expand-response-params>   /nix/store/i3ibpx67yncp4w4mpkf5pwvjjsd0aqln-bootstrap-tools/lib/gcc/x86_64-unknown-linux-gnu/8.3.0/include-fixed
expand-response-params>   -B/nix/store/pq1sjlnlg3jx7jcnlyjmazhm3in5nh5f-bootstrap-stage3-gcc-wrapper-/bin/
expand-response-params>   -frandom-seed=w18da4hqv9
expand-response-params>   -Wl\,-rpath
expand-response-params>   -Wl\,/nix/store/w18da4hqv9nlcxn0pyvmx58ilgdy196p-expand-response-params/lib64
expand-response-params>   -Wl\,-rpath
expand-response-params>   -Wl\,/nix/store/w18da4hqv9nlcxn0pyvmx58ilgdy196p-expand-response-params/lib
expand-response-params>   -L/nix/store/bf7l4af8rrfy1s17af4j2az0khlp5rw0-glibc-2.35/lib
expand-response-params>   -L/nix/store/i3ibpx67yncp4w4mpkf5pwvjjsd0aqln-bootstrap-tools/lib
```

**Comment by trofi** [comment]
Proposed possible fix: https://github.com/NixOS/nixpkgs/pull/158047

**Comment by trofi** [comment]
How I understand the breakage:

1. WRT libraries `stdenv` consists of `libc-so+crt`, `cc-libs+crt`, `gmp-mpc-mpfr-libs-so` components (and search paths to them)
2. Initial `stdenv` consists of prebuild `bootstrapTools` for all 3 components.
3. There are a few intermediate `stdenv` states during bootstrap to gradually switch from `bootstrapTools` prebuilts to `nixpkgs`-provided packages: https://github.com/NixOS/nixpkgs/blob/a5851f18c5c87348d86b36741290201eb43258ac/pkgs/stdenv/linux/default.nix#L137-L430
4. each next `stdenv` stage tries to change only one aspect of `cc-libs/libc-so/cc-libs/gmp-mpc-mpfr-libs-so` by tweaking `-B/-L/-rpath` flags.

The problem is that initial `bootstrapTools` archive has both `libc-so+crt` and `cc-libs` conflated into `${bootstrapTools}/lib`. This makes it hard to swap just `libc-so+crt` out if `${bootstrapTools}/lib` is already the first in the `-L/-B` search path: it always pulls both `libc-so+crt` and `cc-libs+crt`.

My understanding of the bootstrap sequence:
- stage0: use `libc-so+crt`, `cc-libs+crt`, `other-libs-so` from bootstrap
- stage2: build `libc-so+crt`
- stage3: build `gmp-mpc-mpfr-libs-so`
- stage4: build `cc-libs+crt`

I think if we were to unconflate all 3 paths early by splitting `bootstrapTools` into 3 sets of directories it would be easier to reason about the swapping effect of the components.

Otherwise we can try to fix and maintain `-L` and `-B` order to always follow:
- `-L/-B libc-so+crt`
- `-L/-B gmp-mpc-mpfr-libs-so`
- `-L/-B cc-libs+crt`
It sounds fragile to maintain across all the `nix-support/libc-cflags` / `nix-support/libc-crt1-cflags` / `nix-support/cc-cflags-before`

**Comment by trofi** [comment]
Fixed ordering with proposed https://github.com/NixOS/nixpkgs/pull/158047. It does not fix glibc-2.35 yet, but I think it's an improvement.

**Comment by trofi** [comment]
> Fixed ordering with proposed #158047. It does not fix glibc-2.35 yet, but I think it's an improvement.

Found the minor bug in my `glibc-2.35` update. With v2 the fix is enough to get `nixpkgs` bootstrapped with https://github.com/NixOS/nixpkgs/pull/158047.
